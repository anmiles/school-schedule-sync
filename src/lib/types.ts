import { z } from 'zod';

const timeSchema = z.object({
	startTime: z.string().regex(/^\d{2}:\d{2}$/),
	length   : z.number(),
});

const eventSchema = timeSchema.extend({
	name: z.string(),
});

const lessonSchema  = z.string();
const sectionSchema = eventSchema;

const lessonCalendarSchema = z.object({
	name       : z.string(),
	type       : z.literal('lessons'),
	days       : z.array(z.array(lessonSchema)).min(5).max(7),
	lessonTimes: z.array(timeSchema).min(5).max(7).optional(),
});

export type LessonCalendar = z.infer<typeof lessonCalendarSchema>;

const sectionCalendarSchema = z.object({
	name: z.string(),
	type: z.literal('sections'),
	days: z.array(z.array(sectionSchema)).min(5).max(7),
});

export type SectionCalendar = z.infer<typeof sectionCalendarSchema>;

export const scheduleSchema = z.object({
	defaults: z.object({
		lessonTimes: z.array(timeSchema).optional(),
	}),
	calendars: z.array(z.union([
		lessonCalendarSchema,
		sectionCalendarSchema,
	])),
});

export type Schedule = z.infer<typeof scheduleSchema>;
