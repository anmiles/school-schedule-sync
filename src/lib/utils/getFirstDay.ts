export function getFirstDay(weekDay: number, year: number): Date {
	const startDate = new Date(year, 8, 1);
	const startDay  = startDate.getDay();
	const diff      = weekDay - startDay;
	startDate.setDate((diff + 8) % 7 + 1);
	return startDate;
}
