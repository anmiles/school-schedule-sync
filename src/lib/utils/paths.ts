import path from 'path';

const dirPaths = {
	config: 'config',
};

export function getScheduleFile(): string {
	return path.join(dirPaths.config, 'schedule.json');
}

export function getSampleScheduleFile(): string {
	return path.join(dirPaths.config, 'schedule.sample.json');
}
