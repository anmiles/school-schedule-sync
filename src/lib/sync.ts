import fs from 'fs';

import { getAPI } from '@anmiles/google-api-wrapper';
import { info, log } from '@anmiles/logger';
import { validate } from '@anmiles/zod-tools';
import { calendar as api } from 'googleapis/build/src/apis/calendar';
import '@anmiles/prototypes';

import { scheduleSchema  } from './types';
import { getFirstDay } from './utils/getFirstDay';
import { getScheduleFile } from './utils/paths';

const fullScopes = [
	'https://www.googleapis.com/auth/calendar.calendars.readonly',
	'https://www.googleapis.com/auth/calendar.calendarlist.readonly',
	'https://www.googleapis.com/auth/calendar.events.owned',
];

const YEAR     = new Date().getFullYear();
const TIMEZONE = Intl.DateTimeFormat().resolvedOptions().timeZone;

export async function sync(profile: string): Promise<void> {

	const scheduleJSON = fs.readJSON(getScheduleFile());
	const schedule     = validate(scheduleJSON, scheduleSchema);

	const calendarAPI  = await getAPI((auth) => api({ version: 'v3', auth }), profile, { temporary: true, scopes: fullScopes });
	const allCalendars = await calendarAPI.getItems((api) => api.calendarList, {}, { hideProgress: true });

	if (allCalendars.length === 0) {
		throw new Error(`There are no available calendars for profile '${profile}'`);
	}

	for (const calendar of schedule.calendars) {
		info(`Calendar '${calendar.name}'...`);
		const calendarId = allCalendars.find((c) => c.summary === calendar.name)?.id;

		if (!calendarId) {
			throw new Error(`Unknown calendar '${calendar.name}' for profile '${profile}'`);
		}

		log('Clearing calendar...');
		const allEvents = await calendarAPI.getItems((api) => api.events, { calendarId }, { hideProgress: true });

		await allEvents.forEachAsync(async (event) => {
			log(`    ${event.summary}`);
			await calendarAPI.api.events.delete({ calendarId, eventId: event.id ?? '' });
		});

		log('Creating events...');

		const sectionsOrLessons = calendar.type === 'sections'
			? calendar.days
			: calendar.days.map((day) => day.map((name, index) => {
					const time = calendar.lessonTimes?.[index] ?? schedule.defaults.lessonTimes?.[index];

					if (!time) {
						throw new Error(`Cannot find time described for lesson #${index + 1} in calendar '${calendar.name}'`);
					}

					return { name, ...time };
				}));

		await sectionsOrLessons.forEachAsync(async (events, weekDay) => {
			const firstDay = getFirstDay(weekDay, YEAR);
			log(`  #${weekDay + 1}`);

			await events.forEachAsync(async (event) => {
				const [ startHour, startMinute ] = event.startTime.split(':').map((str) => parseInt(str)).toTuple(2);

				const startDate = new Date(firstDay);
				startDate.setHours(startHour);
				startDate.setMinutes(startMinute);

				const endDate = new Date(startDate);
				endDate.setMinutes(endDate.getMinutes() + event.length);

				const dayAbbr = firstDay.toLocaleDateString('en-US', { weekday: 'short' }).slice(0, 2).toUpperCase();

				log(`    ${event.startTime} ${event.name}`);

				await calendarAPI.api.events.insert({
					calendarId,
					requestBody: {
						start     : { dateTime: startDate.toISOString(), timeZone: TIMEZONE },
						end       : { dateTime: endDate.toISOString(), timeZone: TIMEZONE },
						summary   : event.name,
						recurrence: [ `RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=${YEAR + 1}0525T000000Z;BYDAY=${dayAbbr}` ],
					},
				}, {});
			});
		});
	}
}
