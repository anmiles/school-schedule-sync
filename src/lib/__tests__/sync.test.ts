import fs from 'fs';

import { validate } from '@anmiles/express-tools';
import { getAPI } from '@anmiles/google-api-wrapper';
import type GoogleApis from 'googleapis';
// eslint-disable-next-line camelcase
import type { calendar_v3, MethodOptions } from 'googleapis/build/src/apis/calendar';
import { calendar } from 'googleapis/build/src/apis/calendar';
import '@anmiles/jest-extensions';
import mockFs from 'mock-fs';

import { sync } from '../sync';
import {   scheduleSchema } from '../types';
import type { LessonCalendar, Schedule } from '../types';
import { getSampleScheduleFile, getScheduleFile } from '../utils/paths';

jest.mock('@anmiles/google-api-wrapper');
jest.mock('@anmiles/logger');
jest.mock('googleapis/build/src/apis/calendar');

const profile = 'username';

const calendarApis = {
	calendarList: 'calendarList',
	events      : 'events',
} as const;

function mockGetItems(selectAPI: ((api: typeof calendarApis)=> typeof calendarApis[keyof typeof calendarApis])): typeof calendars | typeof events {
	switch (selectAPI(calendarApis)) {
		case calendarApis.calendarList: return calendars;
		case calendarApis.events: return events;
	}
}

const getItems = jest.mocked(jest.fn().mockImplementation(mockGetItems));

const api = {
	events: {
		insert: jest.fn(),
		delete: jest.fn(),
	},
};

// eslint-disable-next-line @typescript-eslint/require-await -- allow partial mock
const getAPIMock = jest.fn().mockImplementation(async () => ({ getItems, api }));

const auth = { kind: 'auth' };

let calendars: Array<{ id?: string | null | undefined; summary?: string; description?: string; hidden?: boolean }>;
let events: Array<{ id?: string | null | undefined; summary?: string; organizer?: { email?: string; displayName?: string; self?: boolean } }>;

jest.mocked(getAPI).mockImplementation((...args: unknown[]) => getAPIMock(...args));

// eslint-disable-next-line @typescript-eslint/no-unsafe-type-assertion
jest.mocked(calendar).mockReturnValue(calendarApis as unknown as GoogleApis.calendar_v3.Calendar);

const scheduleJSON = fs.readJSON(getSampleScheduleFile());
const schedule     = validate(scheduleJSON, scheduleSchema);

beforeEach(() => {
	mockFs({
		[getScheduleFile()]: JSON.stringify(schedule),
	});

	calendars = [
		{ id: 'lessons-1th', summary: 'First grade lessons' },
		{ id: 'lessons-hs', summary: 'High school lessons' },
		{ id: 'sections-1th', summary: 'First grade sections' },
	];

	events = [
		{ id: 'id1' },
		{ },
		{ id: 'id3' },
	];
});

afterAll(() => {
	mockFs.restore();
});

const fullScopes = [
	'https://www.googleapis.com/auth/calendar.calendars.readonly',
	'https://www.googleapis.com/auth/calendar.calendarlist.readonly',
	'https://www.googleapis.com/auth/calendar.events.owned',
];

describe('src/lib/sync', () => {
	describe('sync', () => {
		it('should get calendar API', async () => {
			await sync(profile);

			expect(getAPIMock).toHaveBeenCalledWith(expect.toBeFunction([ auth ], calendarApis), profile, { temporary: true, scopes: fullScopes });
			expect(calendar).toHaveBeenCalledWith({ version: 'v3', auth });
		});

		it('should get all calendars without showing progress', async () => {
			await sync(profile);

			expect(getItems).toHaveBeenCalledWith(expect.anything(), {}, { hideProgress: true });
			expect(getItems.mock.calls[0]?.[0](calendarApis)).toEqual(calendarApis.calendarList);
		});

		it('should throw if there are no available calendars', async () => {
			calendars = [];

			const promise = async (): Promise<unknown> => sync(profile);

			await expect(promise).rejects.toEqual(new Error(`There are no available calendars for profile '${profile}'`));
		});

		it('should throw if there are no matching calendars', async () => {
			const savedSchedule = fs.readJSON<Schedule>(getScheduleFile());
			// eslint-disable-next-line @typescript-eslint/no-unsafe-type-assertion
			const savedCalendar = savedSchedule.calendars[0] as LessonCalendar;

			savedCalendar.name         = 'random calendar name';
			savedSchedule.calendars[0] = savedCalendar;

			mockFs({
				[getScheduleFile()]: JSON.stringify(savedSchedule),
			});

			const promise = async (): Promise<unknown> => sync(profile);

			await expect(promise).rejects.toEqual(new Error(`Unknown calendar 'random calendar name' for profile '${profile}'`));
		});

		it('should throw if no time described for a lesson', async () => {
			const savedSchedule = fs.readJSON<Schedule>(getScheduleFile());
			// eslint-disable-next-line @typescript-eslint/no-unsafe-type-assertion
			const savedCalendar = savedSchedule.calendars[0] as LessonCalendar;

			savedCalendar.days[1]!.push('6th lesson');
			savedSchedule.calendars[0] = savedCalendar;

			mockFs({
				[getScheduleFile()]: JSON.stringify(savedSchedule),
			});

			const promise = async (): Promise<unknown> => sync(profile);

			await expect(promise).rejects.toEqual(new Error('Cannot find time described for lesson #6 in calendar \'First grade lessons\''));
		});

		it('should get events for calendars without showing progress', async () => {
			await sync(profile);

			expect(getItems).toHaveBeenCalledTimes(4);

			expect(getItems).toHaveBeenCalledWith(
				expect.toBeFunction([ calendarApis ], calendarApis.calendarList),
				{ },
				{ hideProgress: true },
			);

			calendars.forEach((calendar) => {
				expect(getItems).toHaveBeenCalledWith(
					expect.toBeFunction([ calendarApis ], calendarApis.events),
					{ calendarId: calendar.id },
					{ hideProgress: true },
				);
			});
		});

		it('should delete all existing events in calendars', async () => {
			await sync(profile);

			expect(api.events.delete).toHaveBeenCalledTimes(9);

			calendars.forEach((calendar) => {
				expect(api.events.delete).toHaveBeenCalledWith({ calendarId: calendar.id, eventId: 'id1' });
				expect(api.events.delete).toHaveBeenCalledWith({ calendarId: calendar.id, eventId: '' });
				expect(api.events.delete).toHaveBeenCalledWith({ calendarId: calendar.id, eventId: 'id3' });
			});
		});

		it('should create all events for schedule', async () => {
			await sync(profile);

			expect(api.events.insert.mock.calls.map(stringifyInsertCall).join('\n')).toMatchSnapshot();
		});
	});
});

// eslint-disable-next-line camelcase
function stringifyInsertCall(insertCall: [calendar_v3.Params$Resource$Events$Insert, MethodOptions]): string {
	return [
		insertCall[0].calendarId,
		JSON.stringify(insertCall[1]),
		insertCall[0].requestBody?.start?.dateTime,
		insertCall[0].requestBody?.start?.timeZone,
		insertCall[0].requestBody?.end?.dateTime,
		insertCall[0].requestBody?.end?.timeZone,
		insertCall[0].requestBody?.recurrence,
		insertCall[0].requestBody?.summary,
	].join('\t');
}
