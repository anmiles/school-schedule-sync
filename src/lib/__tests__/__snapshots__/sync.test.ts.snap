// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`src/lib/sync sync should create all events for schedule 1`] = `
"lessons-1th	{}	2025-09-01T05:00:00.000Z	UTC	2025-09-01T05:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=MO	Reading
lessons-1th	{}	2025-09-01T06:00:00.000Z	UTC	2025-09-01T06:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=MO	Writing
lessons-1th	{}	2025-09-01T07:00:00.000Z	UTC	2025-09-01T07:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=MO	Art
lessons-1th	{}	2025-09-01T08:30:00.000Z	UTC	2025-09-01T09:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=MO	Music
lessons-1th	{}	2025-09-02T05:00:00.000Z	UTC	2025-09-02T05:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TU	Math
lessons-1th	{}	2025-09-02T06:00:00.000Z	UTC	2025-09-02T06:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TU	Story Time
lessons-1th	{}	2025-09-02T07:00:00.000Z	UTC	2025-09-02T07:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TU	Physical Education
lessons-1th	{}	2025-09-02T08:30:00.000Z	UTC	2025-09-02T09:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TU	Drawing
lessons-1th	{}	2025-09-02T09:30:00.000Z	UTC	2025-09-02T10:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TU	Crafts
lessons-1th	{}	2025-09-03T05:00:00.000Z	UTC	2025-09-03T05:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=WE	Nature Study
lessons-1th	{}	2025-09-03T06:00:00.000Z	UTC	2025-09-03T06:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=WE	Handwriting
lessons-1th	{}	2025-09-03T07:00:00.000Z	UTC	2025-09-03T07:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=WE	Singing
lessons-1th	{}	2025-09-03T08:30:00.000Z	UTC	2025-09-03T09:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=WE	Painting
lessons-1th	{}	2025-09-04T05:00:00.000Z	UTC	2025-09-04T05:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TH	Math
lessons-1th	{}	2025-09-04T06:00:00.000Z	UTC	2025-09-04T06:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TH	Reading
lessons-1th	{}	2025-09-04T07:00:00.000Z	UTC	2025-09-04T07:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TH	Game Time
lessons-1th	{}	2025-09-04T08:30:00.000Z	UTC	2025-09-04T09:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TH	Art
lessons-1th	{}	2025-09-04T09:30:00.000Z	UTC	2025-09-04T10:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TH	Story Time
lessons-1th	{}	2025-09-05T05:00:00.000Z	UTC	2025-09-05T05:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=FR	English
lessons-1th	{}	2025-09-05T06:00:00.000Z	UTC	2025-09-05T06:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=FR	Coloring
lessons-1th	{}	2025-09-05T07:00:00.000Z	UTC	2025-09-05T07:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=FR	Physical Education
lessons-1th	{}	2025-09-05T08:30:00.000Z	UTC	2025-09-05T09:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=FR	Story Time
lessons-hs	{}	2025-09-01T11:00:00.000Z	UTC	2025-09-01T11:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=MO	Mathematics
lessons-hs	{}	2025-09-01T12:00:00.000Z	UTC	2025-09-01T12:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=MO	English Literature
lessons-hs	{}	2025-09-01T13:00:00.000Z	UTC	2025-09-01T13:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=MO	Physics
lessons-hs	{}	2025-09-01T14:30:00.000Z	UTC	2025-09-01T15:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=MO	History
lessons-hs	{}	2025-09-01T15:30:00.000Z	UTC	2025-09-01T16:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=MO	Computer Science
lessons-hs	{}	2025-09-02T11:00:00.000Z	UTC	2025-09-02T11:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TU	Chemistry
lessons-hs	{}	2025-09-02T12:00:00.000Z	UTC	2025-09-02T12:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TU	Physical Education
lessons-hs	{}	2025-09-02T13:00:00.000Z	UTC	2025-09-02T13:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TU	English Language
lessons-hs	{}	2025-09-02T14:30:00.000Z	UTC	2025-09-02T15:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TU	Biology
lessons-hs	{}	2025-09-03T11:00:00.000Z	UTC	2025-09-03T11:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=WE	Geography
lessons-hs	{}	2025-09-03T12:00:00.000Z	UTC	2025-09-03T12:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=WE	Mathematics
lessons-hs	{}	2025-09-03T13:00:00.000Z	UTC	2025-09-03T13:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=WE	Economics
lessons-hs	{}	2025-09-03T14:30:00.000Z	UTC	2025-09-03T15:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=WE	Art
lessons-hs	{}	2025-09-03T15:30:00.000Z	UTC	2025-09-03T16:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=WE	Philosophy
lessons-hs	{}	2025-09-04T11:00:00.000Z	UTC	2025-09-04T11:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TH	English Language
lessons-hs	{}	2025-09-04T12:00:00.000Z	UTC	2025-09-04T12:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TH	Physics
lessons-hs	{}	2025-09-04T13:00:00.000Z	UTC	2025-09-04T13:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TH	History
lessons-hs	{}	2025-09-04T14:30:00.000Z	UTC	2025-09-04T15:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=TH	Computer Science
lessons-hs	{}	2025-09-05T11:00:00.000Z	UTC	2025-09-05T11:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=FR	Biology
lessons-hs	{}	2025-09-05T12:00:00.000Z	UTC	2025-09-05T12:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=FR	Mathematics
lessons-hs	{}	2025-09-05T13:00:00.000Z	UTC	2025-09-05T13:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=FR	Chemistry
lessons-hs	{}	2025-09-05T14:30:00.000Z	UTC	2025-09-05T15:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=FR	English Literature
lessons-hs	{}	2025-09-05T15:30:00.000Z	UTC	2025-09-05T16:15:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=FR	Physical Education
sections-1th	{}	2025-09-01T10:00:00.000Z	UTC	2025-09-01T10:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=MO	Lego Club
sections-1th	{}	2025-09-03T10:00:00.000Z	UTC	2025-09-03T10:30:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=WE	Puppet Theater
sections-1th	{}	2025-09-03T11:00:00.000Z	UTC	2025-09-03T11:30:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=WE	Kids Science Club
sections-1th	{}	2025-09-05T10:00:00.000Z	UTC	2025-09-05T10:45:00.000Z	UTC	RRULE:FREQ=WEEKLY;WKST=MO;UNTIL=20260525T000000Z;BYDAY=FR	Dance Class"
`;
